nengo/processes.py

```diff
@@ -260,3 +262,146 @@ def step_presentinput(t):
+    supported. The default interpolation is 'zero', which creates a
+    piecewise function whose values begin at the specified time points.
+    So the above example would be shortcut for::
+
+        def function(t):
+            if t < 0.5:
+                return 0
+            elif t < 0.75
+                return -1
+            elif t < 1:
+                return 0.5
+            else:
+                return 0
+
+    For times before the first specified time, an array of zeros (of
+    the correct length) will be emitted.
+    This means that the above can be simplified to::
+
+        Piecewise({0.5: -1, 0.75: 0.5, 1: 0})
+
+    Parameters
+    ----------
+    data : dict
+        A dictionary mapping times to the values that should be emitted
+        at those times. Times must be numbers (ints or floats), while values
+        can be numbers, lists of numbers, or numpy arrays of numbers.
+    interpolation : str, optional (Default: 'zero')
+        One of 'linear', 'nearest', 'slinear', 'quadratic', 'cubic', or 'zero'.
+        Specifies how to interpolate between times with specified value.
+        'zero' creates a plain piecewise function whose values begin at
+        corresponding time points, while all other options interpolate
+        as described in `scipy.interpolate`.
+
+    Attributes
+    ----------
+    interpolation : str
+        One of 'linear', 'nearest', 'slinear', 'quadratic', 'cubic', or 'zero'.
+        Specifies how to interpolate between times with specified value.
+        'zero' creates a plain piecewise function whose values begin at
+        corresponding time points, while all other options interpolate
+        as described in `scipy.interpolate`.
+    tp : ndarray
+        Specified time points.
+    yp : ndarray
+        Values corresponding to the time points, ``tp``.
+
+    Examples
+    --------
+
+    >>> from nengo.processes import Piecewise
+    >>> process = Piecewise({0.5: 1, 0.75: -1, 1: 0})
+    >>> with nengo.Network() as model:
+    ...     u = nengo.Node(process, size_out=process.default_size_out)
+    ...     up = nengo.Probe(u)
+    >>> with nengo.Simulator(model) as sim:
+    ...     sim.run(1.5)
+    >>> f = sim.data[up]
+    >>> t = sim.trange()
+    >>> f[t == 0.2]
+    array([[ 0.]])
+    >>> f[t == 0.58]
+    """
+
+    tp = NdarrayParam('tp', shape=('*',), optional=True)
+    yp = NdarrayParam('yp', shape=('...',), optional=True)
+    interpolation = EnumParam('interpolation', values=(
+        'zero', 'linear', 'nearest', 'slinear', 'quadratic', 'cubic'))
+
+    def __init__(self, data, interpolation='zero', **kwargs):
+        tp, yp = zip(*data.items())
+        self.tp = tp
+        self.yp = yp
+
+        if self.tp.shape[0] != self.yp.shape[0]:
+            raise ValidationError(
+                "`tp.shape[0]` (%d) must equal `yp.shape[0]` (%d)"
+                % (self.tp.shape[0], self.yp.shape[0]),
+                attr='yp', obj=self)
```

Comment:
There is no way this ValidationError could ever trigger (since a dictionary always has the same number of keys and values).